llvm/lib/Analysis/DependenceAnalysis.cpp

```diff
@@ -3569,6 +3569,123 @@ bool DependenceInfo::invalidate(Function &F, const PreservedAnalyses &PA,
+        // SrcSCEV = ((4 * (sext i8 %n to i64))<nsw> + %A)
+        // DstSCEV = (4 + (4 * (sext i8 %n to i64))<nsw> + %A)
+        // Param = (4 * (sext i8 %n to i64))<nsw>
+        // V = 4 + (4 * (sext i8 %n to i64))<nsw>
+        // Diff = -4, Rem = 0, and so all offsets are multiple of 4.
+        return true;
+      LLVM_DEBUG(dbgs() << "SCEV with different offsets: " << *Param << " - "
+                        << *V << " = " << *Diff << " % " << EltSize << " = "
+                        << Rem << " != 0\n");
+      return false;
+    }
+    // Check if the symbolic difference is a multiple of Size.
+    const SCEV *Val =
+        SE->getConstant(APInt(Diff->getType()->getScalarSizeInBits(), EltSize));
+
+    // Check by using the remainder computation.
+    const SCEV *Remainder = SE->getURemExpr(Diff, Val);
+    if (const SCEVConstant *C = dyn_cast<SCEVConstant>(Remainder))
+      if (C->getValue()->isZero())
+        // For example test/Analysis/DependenceAnalysis/DADelin.ll
+        // SrcSCEV = {{{%A,+,(4 * %m * %o)}<%for.cond1.preheader>,+,(4 * %o)}
+        // DstSCEV = {{{%A,+,(4 * %m * %o)}<%for.cond1.preheader>,+,(4 * %o)}
+        // The strides '(4 * %m * %o)' and '(4 * %o)' are multiple of 4.
+        return true;
+
+    // Check by using the division computation.
+    const SCEV *Q = SE->getUDivExpr(Diff, Val);
+    const SCEV *Product = SE->getMulExpr(Q, Val);
+    if (Diff == Product)
+      return true;
+    LLVM_DEBUG(dbgs() << "SCEV with different offsets:\n"
+                      << *Param << " - " << *V << " = " << *Diff << "\n"
+                      << "Remainder = " << *Remainder << "\n"
+                      << "Q = " << *Q << " Product = " << *Product << "\n");
+    // For example in test/Analysis/DependenceAnalysis/MIVCheckConst.ll
+    // SrcSCEV = {(80640 + (4 * (1 + %n) * %v1) + %A),+,(8 * %v1)}<%bb13>
+    // DstSCEV = {(126720 + (128 * %m) + %A),+,256}<%bb13>
+    // We fail to prove that the offsets 80640 + (4 * (1 + %n) * %v1) and
+    // (8 * %v1) are multiples of 128.
+    // Param = 80640 + (4 * (1 + %n) * %v1)
+    // (80640 + (4 * (1 + %n) * %v1)) - (8 * %v1) =
+    // (80640 + ((-4 + (4 * %n)) * %v1))
+    // Remainder = (zext i7 ((trunc i32 %v1 to i7) *
+    //                       (-4 + (4 * (trunc i32 %n to i7)))) to i32)
+    // Q = ((80640 + ((-4 + (4 * %n)) * %v1)) /u 128)
+    // Product = (128 * ((80640 + ((-4 + (4 * %n)) * %v1)) /u 128))<nuw>
+    return false;
+  };
+
+  // Expressions like "n".
+  if (isa<SCEVUnknown>(V))
+    return checkParamsMultipleOfSize(V, Param);
+
+  // Expressions like "n + 1".
+  if (isa<SCEVAddExpr>(V))
+    return !SCEVExprContains(V, [](const SCEV *S) {
+      return isa<SCEVUnknown>(S);
+    }) || checkParamsMultipleOfSize(V, Param);
+
+  // Expressions like "n * 2".
+  if (isa<SCEVMulExpr>(V))
+    return !SCEVExprContains(V, [](const SCEV *S) {
+      return isa<SCEVUnknown>(S);
+    }) || checkParamsMultipleOfSize(V, Param);
```

Comment:
```suggestion
  if (isa<SCEVAddExpr>(V) || isa<SCEVMulExpr>(V))
    return !SCEVExprContains(V, [](const SCEV *S) {
      return isa<SCEVUnknown>(S);
    }) || checkParamsMultipleOfSize(V, Param);
```

Why must it not contain a SCEVUnknown?